### DISCUSSION TITLE
Validating Arrays in Query Params

### DISCUSSION THREAD

**Discussion by epotter2297** [ecyrbe/zodios#211] [Ideas]
If I have a query parameter like this one:
```js
{
  name: 'filter',
  type: 'Query',
  schema: z.array(z.string()),
},
```
It will fail validation if I pass in an array of length 1. I suppose that the reason it fails makes sense, since you would send in the request like `filter=filter1&filter=filter2...`. I could get around this by making a zod union and then transforming, but then zodios-openapi can't figure out what's happening, and it starts to get a little verbose.

It would be cool if zodios was smart enough to figure out when to do this transform automatically 🙂

**Comment by ecyrbe** [comment]
Hello, 
Can you provider more information ?
Is the validation failing on server side or client side ?

Can you provide a more complete test example to reproduce the issue?
- test setup
-  parameters provided
- expected result

**Comment by ecyrbe** [comment]
@epotter2297 i think i found that you are testing backend. So i know what's wrong and how you should fix it.

test case:
```ts
const api = makeApi([
  {
    method: "get",
    path: "/users",
    alias: "getUsers",
    description: "Get all users",
    parameters: [
      {
        name: "filter",
        type: "Query",
        description: "Filter users by name",
        schema: z.array(z.string()).optional(),
      },
    ],
    response: z.array(userSchema),
  },
```

now when calling this API with just :  
```bash
curl 'http://localhost:3000/api/v1/users?filter=test -H 'accept: application/json'
```
we get the following error:
```json
{
  "context":"query.filter",
  "error":[{
    "code":"invalid_type",
    "expected":"array",
    "received":"string",
    "path":[],
    "message":"Expected array, received string"
  }]
}
```

This is normal, you are calling the API without helping it to understand you mean it to be an array.
This is how you should call it (ie, zodios client does this by default, so if you use zodios client, this works out of the box) :  
```bash
curl 'http://localhost:3000/api/v1/users?filter[]=test -H 'accept: application/json'
#                                               ^ PUT BRACKETS
```
Then there is no errors and zodios correctly parses the parameters as an array

**Reply by ecyrbe** [reply]
I also found out that openapi generation does not send arrays in this format, so a fix for openapi generation is incoming

**Reply by ecyrbe** [reply]
Fixed in `@zodios/openapi` [v10.4.1](https://github.com/ecyrbe/zodios-openapi/releases/tag/v10.4.1)

**Comment by epotter2297** [comment]
Sorry I should have clarified earlier, but yeah you guessed the issue correctly. I do see that with the new `@zodios/openapi` v10.4.1, the `[]` is appended properly, but not when I add a custom `.refine()` like so:
```js
{
  name: 'filter',
  type: 'Query',
  schema: z
    .array(z.string())
    .refine((array) => array.length === new Set(array).size, {
      message: 'Duplicate values are not allowed',
    })
    .optional(),
},
```
Then the `[]` is not appended, probably because the type is now `z.ZodOptional<z.ZodEffects<z.ZodArray<z.ZodString...>>>`

**Reply by ecyrbe** [reply]
Yes, this one is tricky.
i added introspection to zodtype to Guess if it's an arrays, but it looks like it's not sufficient.
I'll take a look into it.

**Reply by ecyrbe** [reply]
@epotter2297 Can you try with z.preprocess instead of refine to see if it works better?

**Reply by epotter2297** [reply]
I don't think that will quite do what I want? With the example above, there wouldn't be a way to fail validation for duplicate values unless I was manually throwing errors in the preprocess step (unless I'm missing something).

**Reply by ecyrbe** [reply]
Yes, you are right.
i found out the cause of this. I'm preparing a fix on both `@zodios/express` and `@zodios/openapi`

**Reply by ecyrbe** [reply]
This should now work on `@zodios/express` v10.4.0 + `@zodios/openapi` v10.4.2

Also note something that maybe i do not higlight enough on the docs :  
- by default on server zod effects are disabled (refine, preprocess, transform) - you can still enable effect with zodios app option `transform`
- indeed when declaring a schema, if you use both the client + server the transform would be done twice.

I'll may change this behaviour soon and propose to declare client / server disctinct zod schema validation. not sure yet how i will do this in details.
